File: resources/[core]/rpg-hud/client/hud.lua
-- ===========================================================================
--  rpg-hud — HUD (player, status, money, paycheck, voice, activity)
--  Event-driven. Un singur thread de polling pentru health, unul pentru voice.
--  Datele "de cont" (username, id, cash, bank, online) vin de la SERVER.
-- ===========================================================================

local hud = {
    started    = false,
    food       = Config.Hud.startNeeds.food,
    water      = Config.Hud.startNeeds.water,
    lastHealth = -1,
    lastVoice  = nil,
}

local function push(kind, value)
    HUD.send({ mod = 'hud', action = kind, value = value })
end

-- ----- feeders (pornesc dupa spawn-ul personajului) -----------------
AddEventHandler('rpg-hud:startFeeders', function()
    if hud.started then return end
    hud.started = true

    -- HEALTH: native client, trimis doar la schimbare
    CreateThread(function()
        while true do
            local ped = PlayerPedId()
            local pct = 0
            if not IsEntityDead(ped) then
                local hp, maxhp = GetEntityHealth(ped), GetEntityMaxHealth(ped)
                if hp > 100 then
                    pct = math.floor((hp - 100) / math.max(1, maxhp - 100) * 100 + 0.5)
                end
            end
            if pct ~= hud.lastHealth then
                hud.lastHealth = pct
                push('health', pct)
            end
            Wait(Config.Hud.healthPollMs)
        end
    end)

    -- NEEDS: mock in dev; altfel doar valorile primite prin evenimente
    push('food', math.floor(hud.food))
    push('water', math.floor(hud.water))
    if Config.Hud.mockNeeds then
        CreateThread(function()
            while true do
                Wait(10000)   -- tick la 10s
                hud.food  = math.max(0, hud.food  - Config.Hud.mockDrainPerMin.food  / 6.0)
                hud.water = math.max(0, hud.water - Config.Hud.mockDrainPerMin.water / 6.0)
                push('food', math.floor(hud.food))
                push('water', math.floor(hud.water))
            end
        end)
    end

    -- VOICE: stare "talking" reala (voice built-in FiveM); restul e API
    CreateThread(function()
        while true do
            local st = hud.lastVoice
            if hud.lastVoice ~= 'muted' then
                st = NetworkIsPlayerTalking(PlayerId()) and 'talking' or 'idle'
            end
            if st ~= hud.lastVoice then
                hud.lastVoice = st
                push('voice', st)
            end
            Wait(Config.Hud.voicePollMs)
        end
    end)

    -- PAYCHECK: seed pentru countdown-ul din NUI (serverul poate suprascrie)
    push('paycheck', { seconds = Config.Hud.paycheckInterval, running = true })
end)

-- ----- API events (framework) ---------------------------------
RegisterNetEvent('hud:updatePlayer',   function(d) push('player', d) end)
RegisterNetEvent('hud:updateOnline',   function(n) push('online', tonumber(n) or 0) end)
RegisterNetEvent('hud:updateMoney',    function(v) push('money', tonumber(v) or 0) end)
RegisterNetEvent('hud:updateBank',     function(v) push('bank', tonumber(v) or 0) end)
RegisterNetEvent('hud:updateHealth',   function(v) push('health', tonumber(v) or 0) end)
RegisterNetEvent('hud:updateFood',     function(v) hud.food  = tonumber(v) or hud.food;  push('food',  math.floor(hud.food)) end)
RegisterNetEvent('hud:updateWater',    function(v) hud.water = tonumber(v) or hud.water; push('water', math.floor(hud.water)) end)
RegisterNetEvent('hud:updatePaycheck', function(d) push('paycheck', d) end)
RegisterNetEvent('hud:updateActivity', function(d) push('activity', d) end)
RegisterNetEvent('hud:clearActivity',  function() push('activityClear', true) end)

-- ----- exports client ---------------------------------------
exports('setFood',      function(v) hud.food = tonumber(v) or hud.food; push('food', math.floor(hud.food)) end)
exports('setWater',     function(v) hud.water = tonumber(v) or hud.water; push('water', math.floor(hud.water)) end)
exports('setVoiceState',function(s) hud.lastVoice = s; push('voice', s) end)
exports('setActivity',  function(d) push('activity', d) end)
exports('clearActivity',function() push('activityClear', true) end)

File: resources/[core]/rpg-hud/html/js/hud.js
/* ==========================================================================
   rpg-hud — HUD (player, money, status bars, paycheck, voice, activity)
   ========================================================================== */
HUD.mods.hud = (function () {
    var $ = HUD.$;

    /* ---- bars ---- */
    function setBar(name, pct) {
        pct = Math.max(0, Math.min(100, Math.round(pct)));
        var b = $('#b-' + name), p = $('#b-' + name + '-p');
        if (b) b.style.width = pct + '%';
        if (p) p.textContent = pct + '%';
        var bar = b && b.closest('.bar');
        if (bar) {
            bar.classList.toggle('low', pct <= 20);
            bar.classList.toggle('mid', pct > 20 && pct <= 45);
        }
    }

    function money(sel, v) {
        var el = $(sel);
        if (el) el.textContent = '$' + Number(v || 0).toLocaleString('en-US');
    }

    /* ---- paycheck countdown (tick local, seed de la client/server) ---- */
    var pc = { seconds: 0, running: false, t: null };
    function renderPc() {
        var m = Math.floor(pc.seconds / 60), s = pc.seconds % 60;
        $('#pc-time').textContent = (m < 10 ? '0' : '') + m + ':' + (s < 10 ? '0' : '') + s;
    }
    function startPaycheck(d) {
        d = d || {};
        pc.seconds = Math.max(0, Math.floor(d.seconds || 0));
        pc.running = d.running !== false;
        if (pc.t) clearInterval(pc.t);
        renderPc();
        pc.t = setInterval(function () {
            if (pc.running && pc.seconds > 0) { pc.seconds--; renderPc(); }
        }, 1000);
    }

    /* ---- activity / status container (reutilizabil) ---- */
    var act = { timer: 0, t: null };
    function fmtDur(s) {
        var h = Math.floor(s / 3600), m = Math.floor((s % 3600) / 60), x = s % 60;
        return (h > 0 ? (h < 10 ? '0' : '') + h + ':' : '') +
            (m < 10 ? '0' : '') + m + ':' + (x < 10 ? '0' : '') + x;
    }
    function setActivity(d) {
        if (!d) return clearActivity();
        $('#hud-activity').hidden = false;
        $('#a-title').textContent = d.title || 'ACTIVITATE';
        if (act.t) { clearInterval(act.t); act.t = null; }

        if (typeof d.timer === 'number') {
            act.timer = Math.floor(d.timer);
            var tick = function () {
                $('#a-sub').textContent = fmtDur(Math.max(0, act.timer));
                if (act.timer <= 0) { clearInterval(act.t); act.t = null; }
                else act.timer--;
            };
            tick();
            act.t = setInterval(tick, 1000);
        } else {
            $('#a-sub').textContent = d.text || '';
        }

        var barWrap = $('#hud-activity .a-bar');
        if (typeof d.progress === 'number') {
            barWrap.hidden = false;
            $('#a-fill').style.width = Math.round(Math.max(0, Math.min(1, d.progress)) * 100) + '%';
        } else {
            barWrap.hidden = true;
        }
    }
    function clearActivity() {
        $('#hud-activity').hidden = true;
        if (act.t) { clearInterval(act.t); act.t = null; }
    }

    return {
        on: function (action, value) {
            switch (action) {
                case 'health': setBar('health', value); break;
                case 'food': setBar('food', value); break;
                case 'water': setBar('water', value); break;
                case 'player':
                    $('#p-name').textContent = (value && value.username) || '—';
                    $('#p-id').textContent = '#' + (value && value.id != null ? value.id : '—');
                    break;
                case 'online': $('#p-online').textContent = value; break;
                case 'money': money('#m-cash', value); break;
                case 'bank': money('#m-bank', value); break;
                case 'paycheck': startPaycheck(value); break;
                case 'voice': $('#voice').className = 'voice ' + (value || 'idle'); break;
                case 'activity': setActivity(value); break;
                case 'activityClear': clearActivity(); break;
            }
        },
    };
})();
